server/src/main/java/org/apache/cassandra/sidecar/acl/authorization/CachedAuthorizationHandler.java

```diff
@@ -0,0 +1,201 @@
+    private final AuthorizationParameterValidateHandler authZParameterValidateHandler;
+    private final AdminIdentityResolver adminIdentityResolver;
+    private final AuthMetrics authMetrics;
+    private final AsyncCache<AuthorizationCacheKey, Boolean> authorizationCache;
+
+    // This is overridden since Vert.x does not expose this
+    private BiConsumer<RoutingContext, AuthorizationContext> variableHandler;
+
+    public CachedAuthorizationHandler(AccessControlConfiguration accessControlConfiguration,
+                                      AuthorizationParameterValidateHandler authZParameterValidateHandler,
+                                      AdminIdentityResolver adminIdentityResolver,
+                                      Authorization authorization,
+                                      SidecarMetrics sidecarMetrics,
+                                      AsyncCache<AuthorizationCacheKey, Boolean> authorizationCache)
+    {
+        this(HANDLER_ID_GEN.getAndIncrement(), accessControlConfiguration, authZParameterValidateHandler,
+             adminIdentityResolver, authorization, sidecarMetrics, authorizationCache);
+    }
+
+    @VisibleForTesting
+    public CachedAuthorizationHandler(int handlerId,
+                                      AccessControlConfiguration accessControlConfiguration,
+                                      AuthorizationParameterValidateHandler authZParameterValidateHandler,
+                                      AdminIdentityResolver adminIdentityResolver,
+                                      Authorization authorization,
+                                      SidecarMetrics sidecarMetrics,
+                                      AsyncCache<AuthorizationCacheKey, Boolean> authorizationCache)
+    {
+        super(authorization);
+        this.handlerId = handlerId;
+        this.accessControlConfiguration = accessControlConfiguration;
+        this.authZParameterValidateHandler = authZParameterValidateHandler;
+        this.adminIdentityResolver = adminIdentityResolver;
+        this.authMetrics = sidecarMetrics.server().auth();
+        this.authorizationCache = authorizationCache;
+    }
+
+    @Override
+    public void handle(RoutingContext ctx)
+    {
+        long startTimeNanos = System.nanoTime();
+        authZParameterValidateHandler.handle(ctx);
+        if (ctx.failed()) // failed due to validation
+        {
+            return;
+        }
+
+        User user = ctx.user();
+        AuthorizationContext authorizationContext = AuthorizationContext.create(user);
+        if (this.variableHandler != null)
+        {
+            this.variableHandler.accept(ctx, authorizationContext);
+        }
+
+        AtomicBoolean ctxNextCalled = new AtomicBoolean(false);
+        AuthorizationCacheKey key = AuthorizationCacheKey.create(handlerId, authorizationContext);
+        CompletableFuture<Boolean> authorizationFuture = checkAuthorization(key, ctx, ctxNextCalled, startTimeNanos);
```

Comment:
`checkAuthorization` should not return a CompletableFuture. The code base is using vertx and we should consistently use vertx Futures instead here. Vertx provides ways to convert a CompletableFuture to Future, `Future.fromCompletionStage`.

If we start having an inconsistent code base it will become harder for new contributors to understand when to use one or the other. Having a single way of doing things will simplify this.

```suggestion
        Future<Boolean> authorizationFuture = checkAuthorization(key, ctx, ctxNextCalled, startTimeNanos);
```

Reply:
We have lot of usage of `CompletableFuture` in Sidecar project in client subproject.

Reply:
the reason for the client to use CompletableFuture is because the core project client project does not have any vertx dependencies